File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
//const arr = input[1].split("");
const map = input.slice(1).map((_) => _.trim().split(" ").map(Number));

function main() {
  const edges = Array.from({ length: N + 1 }, () => []);
  map.map(([u, v]) => {
    edges[u].push(v);
    edges[v].push(u);
  });

  let answer = 0;

  const visited = Array.from({ length: N + 1 }, () => false);
  for (let i = 1; i < edges.length; i++) {
    if (visited[i]) continue;

    visited[i] = true;

    const queue = edges[i];

    while (queue.length > 0) {
      const cur = queue.shift();

      edges[cur].forEach((vertex) => {
        if (visited[vertex]) return;
        queue.push(vertex);
        visited[vertex] = true;
      });

      visited[cur] = true;
    }

    answer++;
  }

  console.log(answer);
}

main();

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

   // freopen("input.txt", "r", stdin);

    int N, M;
    cin >> N >> M;

    vector<int> edges[N+1]; 
    for(int i=0;i<M;i++) {
        int u, v;
        cin >> u >> v;

        edges[u].push_back(v);
        edges[v].push_back(u);
    }

    int answer = 0;

    bool visited[N+1] = { false };

    for(int i=1; i<N+1; i++) {
        if(visited[i]) continue;
        visited[i] = true;

        queue<int> q;

        for(int j=0;j<edges[i].size(); j++) {
            int vertex = edges[i][j];
            q.push(vertex);
            visited[vertex] = true;
        }

        while(!q.empty()) {
            int curV = q.front();
            q.pop();

            for(int k=0;k<edges[curV].size(); k++) {
                int nextV = edges[curV][k];
                if(visited[nextV]) continue;
                q.push(nextV);
                visited[nextV] = true;
            }
        }

        answer++;
    }

    cout << answer;
    return 0;
}